lib/ansible/plugins/inventory/__init__.py

```diff
@@ -415,7 +415,11 @@ def _add_host_to_keyed_groups(self, keys, variables, host, strict=False):
                             raise AnsibleParserError("Invalid group name format, expected a string or a list of them or dictionary, got: %s" % type(key))
 
                         for bare_name in new_raw_group_names:
-                            gname = to_safe_group_name('%s%s%s' % (prefix, sep, bare_name))
+                            if keyed.get('unsafe', False):
+                                gname = '%s%s%s' % (prefix, sep, bare_name)
```

Comment:
Right now I get conflicting warnings about the group replacing invalid characters and then that the group doesn't exist (and then inventory fails to parse) when not setting the global toggle but using `unsafe`. `force` and `silent` should probably become parameters of that function on line 42 so you can call that instead.